### PR TITLE
[Feat] Support InternVL sft training in xtuner lite

### PR DESCRIPTION
**Pull request by C1rN09**
# [Feat] Enable SFT Training for InternVL Models in XTuner Lite

This PR introduces foundational support for Supervised Fine-Tuning (SFT) of InternVL models in XTuner Lite, implementing a basic parallel training strategy while maintaining compatibility with the latest XTuner development branch. The implementation **prioritizes correctness and simplicity**, establishing a baseline for future optimizations (e.g., advanced parallel strategies, PyTorch 2.0 compilation).

## Implementation Highlights

- **Parallel Strategy Framework**:
  - Language Model (LLaMA/Qwen): Utilizes existing FSDP2 + TP + SP + `torch.compile` paradigms
  - Vision Transformer (ViT): Implements pure FSDP2 with input chunking and output all-gather
- **System Integration**:
  - Updated InternVL2 dataset adaption for XTuner development branch

## Integration Example
```python
internvl = AutoPatch.from_causal_lm(
    internvl,
    fsdp_config=FSDPConfig(
        tp_size=args.tp_size,
        sp_size=1  # SP parallelism currently constrained
    ),
)
internvl.fully_shard()
```

## Compatibility Considerations

1. **API Modifications**:
   - Added optional `module2name` and `checkpoint_loader` parameters to `fully_shard`
   - Maintains backward compatibility through default `None` values

2. **Embedding Layer Handling**:
   - Automatic resizing to `world_size` multiples during training
   - Original dimensions restored pre-checkpointing
   - *Note:* Potential accuracy impact when original vocab size ≠ `world_size` multiple

3. **Embedding Layer Sharding**:
   - Implemented separate `FSDPParamGroup` for language model embeddings
   - Observed accuracy degradation on specific settings (PyTorch 2.7 and inter-node comm, i.e. >= 2 nodes), but PyTorch 2.5.1 works fine
   - Further investigation required

### CODE REVIEW

File: xtuner/_lite/patches/llama.py [hunk 1]
```diff
@@ -462,6 +478,13 @@ def fully_shard(self, fsdp_config: FSDPConfig) -> None:
                 self.casual_tp_plan,
             )
 
+        fully_shard(
```
Comment (hhaAndroid) [inline xtuner/_lite/patches/llama.py]: 这个地方单独切分是有啥特殊原因？

File: xtuner/_lite/patches/internvl_chat.py [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+        )
+
+        mp_policy = MixedPrecisionPolicy(
+            param_dtype=fsdp_config.param_dtype,
+            reduce_dtype=fsdp_config.reduce_dtype,
+        )
+
+        vision_model = self.patched_model.vision_model
+        # compiled_layers: List[nn.Module] = []
+        for layer_idx, layer in enumerate(vision_model.encoder.layers):
```
Comment (hhaAndroid) [inline xtuner/_lite/patches/internvl_chat.py:191]: 感觉 vit 和 llm 可以有不同的切分规则和参数，但是现在只能用同一套fsdp_config，不是很友好

File: xtuner/_lite/patches/internvl_chat.py [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+            for layer_cur, layer_next in zip(
+                vision_model.encoder.layers[:-1],
+                vision_model.encoder.layers[1:],
+            ):
+                layer_cur.set_modules_to_forward_prefetch([layer_next])
+
+        vision_model.embeddings.apply(param_init_fn)
+        self.patched_model.mlp1.apply(param_init_fn)
+
+        fully_shard(
```
Comment (hhaAndroid) [inline xtuner/_lite/patches/internvl_chat.py:227]: vit 暂时是不支持 tp 是吧？

File: xtuner/_lite/patches/internvl_chat.py [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+        image_flags: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[List[torch.FloatTensor]] = None,
+        labels: Optional[torch.LongTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        tp_mesh: Optional[DeviceMesh] = None,
+        **kwargs,
+    ) -> Union[Tuple, CausalLMOutputWithPast]:
```
Comment (hhaAndroid) [inline xtuner/_lite/patches/internvl_chat.py:251]: 支持 sp

File: xtuner/_lite/patches/llama.py (hunk 1 above)
Comment (C1rN09) [reply xtuner/_lite/patches/llama.py]: 在 ViT 里面需要直接调用 Embedding 层的 forward。如果不作为一个单独的 FSDP unit，那么调用 forward 的时候参数没有被 Shard -> Replicate

File: xtuner/_lite/patches/internvl_chat.py (hunk 2 above)
Comment (C1rN09) [reply xtuner/_lite/patches/internvl_chat.py:191]: 为了接口一致性，暂时使用了这种方式。但是 vit 的部分暂时是纯 FSDP 实现的

File: xtuner/_lite/patches/internvl_chat.py (hunk 3 above)
Comment (C1rN09) [reply xtuner/_lite/patches/internvl_chat.py:227]: 不支持。有 3 个主要的阻碍点：
1. PyTorch Convolution 层的 DTensor 实现[有些问题](https://github.com/pytorch/pytorch/issues/142058)，至少在 pytorch 2.5.1 还未修复
2. InternVL attention 中的 qkv Liner 输出的布局是 (3, num_head, head_dim），直接使用 ColwiseParallel 对 qkv 做 Shard(1) 操作会导致输出不符合预期。这里需要对 attn 计算部分进行 patch 修改
3. InternVL 的 attention 输入输出的 seq_len 是 patch_size^2 + 1，是一个奇数，不能被 tp_size 整除。如果不手动 padding，由于 [PyTorch 的 bug](https://github.com/pytorch/pytorch/issues/143748)会导致运行报错；如果手动 padding，需要添加很多的 patch function，把 cu_seq_lens 等参数一层一层地传进去

考虑到目前的 ViT 部分普遍比较小，并行策略带来的速度影响不大；同时为了保持这个 PR 的正确性和最简实现，因此 ViT 的部分只使用纯 FSDP 来实现。tp/sp 等特性可以后续再添加

File: xtuner/_lite/patches/internvl_chat.py (hunk 4 above)
Comment (C1rN09) [reply xtuner/_lite/patches/internvl_chat.py:251]: 暂时不打算在此 PR 中支持。目前会在 `forward` 阶段 `AssertionError` 提醒用户
